components/places/src/match_impl.rs

```diff
@@ -0,0 +1,483 @@
+        // encoding of the corresponding special case character, and look for it
+        // in the loop below.  For other characters we fall back to 0xff, which
+        // is not a valid UTF-8 byte.
+        let target = dubious_to_ascii_lower(search_for as u8);
+        let special = if target == b'i' { 0xc4u8 }
+                      else if target == b'k' { 0xe2u8 }
+                      else { 0xffu8 };
+        // Note: rustc doesn't do great at all on the more idiomatic
+        // implementation of this (or below), but it does okay for this.
+        let mut ci = 0;
+        while ci < search_bytes.len() {
+            let cur = search_bytes[ci];
+            if dubious_to_ascii_lower(cur) == target || cur == special {
+                return Some(ci);
+            }
+            ci += 1;
+        }
+    } else {
+        let mut ci = 0;
+        while ci < search_bytes.len() {
+            let cur = search_bytes[ci];
+            if (cur & 0x80) != 0 {
+                return Some(ci);
+            }
+            ci += 1;
+        }
+    }
+    None
+}
+
+#[inline(always)]
+fn is_ascii_lower_alpha(c: u8) -> bool {
+    // Equivalent to (but fewer operations than) `b'a' <= c && c <= b'z'`
+    c.wrapping_sub(b'a') <= (b'z' - b'a')
+}
+
+/// port of isOnBoundary from gecko places.
+///
+/// > Check whether a character position is on a word boundary of a UTF-8 string
+/// > (rather than within a word).  We define "within word" to be any position
+/// > between [a-zA-Z] and [a-z] -- this lets us match CamelCase words.
+/// > TODO: support non-latin alphabets.
+#[inline(always)]
+fn is_on_boundary(text: &str, index: usize) -> bool {
+    let bytes = text.as_bytes();
+    if is_ascii_lower_alpha(bytes[index]) {
+        let prev_lower = dubious_to_ascii_lower(bytes[index - 1]);
+        !is_ascii_lower_alpha(prev_lower)
+    } else {
+        true
+    }
+}
+
+/// Returns true if `source` starts with `token` ignoring case.
+///
+/// Loose port of stringMatch from places, that does more correct case folding.
+/// (If this turns out to be a perf issue we can always address it then)
+#[inline]
+fn string_match(token: &str, source: &str) -> bool {
+    if source.len() < token.len() {
+        return false;
+    }
+    let mut ti = token.chars().default_case_fold();
+    let mut si = source.chars().default_case_fold();
+    let mut i = 0;
+    loop {
+        match (ti.next(), si.next()) {
+            (None, _) => return true,
+            (Some(_), None) => return false,
+            (Some(x), Some(y)) => if x != y { return false },
+        }
+        i += 1;
+        if i > token.len() + source.len() {
+            panic!("Bug: {:?} {:?}", token, source);
+        }
+    }
+}
+
+/// This performs single-codepoint case folding. It will do the wrong thing
+/// for characters which .
+#[inline]
+fn char_to_lower_single(c: char) -> char {
+    c.to_lowercase().next().unwrap()
+}
+
+#[inline]
+fn next_codepoint_lower(s: &str) -> (usize, char) {
```

Comment:
So this returns the index of the next character, and the lowercase _current_ (first) character? Mind adding a comment?